src/platforms/web/compiler/directives/model.js

```diff
@@ -163,12 +163,12 @@ function genSelect (
   }
 
   const number = modifiers && modifiers.number
-  const assignment = `Array.prototype.filter` +
+  const selectedVal = `Array.prototype.filter` +
     `.call($event.target.options,function(o){return o.selected})` +
     `.map(function(o){var val = "_value" in o ? o._value : o.value;` +
-    `return ${number ? '_n(val)' : 'val'}})` +
-    (el.attrsMap.multiple == null ? '[0]' : '')
+    `return ${number ? '_n(val)' : 'val'}})`
 
+  const assignment = `$event.target.multiple ? ${selectedVal}: ${selectedVal}[0]`
```

Comment:
This can make the generated code quite verbose since `selectedVal` is repeated twice. We can store it in a runtime temp variable (e.g. `$$selectedVal`)